### ISSUE TITLE
Realm file reached ~460MB in size

### ISSUE THREAD

**Issue by luongvanphuc** [realm/realm-cocoa#2412]
Hi Realm team,

Yesterday, I've been using realm to store data. But this afternoon, the realm file's size reached ~460MB in size with just about 2000 records. I'm doubting that my schema or the index cause this issue, but don't know the root cause and how to fix that. I've attached the file below. Please help me to indicate that my schema is correct defined or not and what is the exact issue.
https://drive.google.com/file/d/0BwRYvsGyoS-uSkdldmJSWXdDVUk/view?usp=sharing

P/S: besides objectID which is primary key, I also define index for "name", "bundleID", "startedTime" and "duration". I'm using realm-objc-0.94.1 for OSX app.

Thanks so much for your time and support, I'm looking forward to hearing from you soon.
Gin.

**Comment by bdash** [comment]
If I use `-[RLMRealm writeCopyToPath:error:]` to write a copy of your Realm to a different path, I see that the copy is only ~275KB. This suggests that your Realm contains a large number of intermediate versions of the data which we've not yet been able to reclaim. This  The [File size & tracking of intermediate versions](https://realm.io/docs/objc/latest/#file-size--tracking-of-intermediate-versions) section in the Realm documentation talks about why this can happen, and what you can do to avoid it.

The general idea is:
  1. When using an `RLMRealm` from a thread with a runloop, such as the main thread, ensure that the runloop is able to iterate normally. In other words: avoid blocking the thread for a long time.

  2. When using an `RLMRealm` from a thread without a runloop, such as Grand Central Dispatch's queues, ensure that the object doesn't live longer than you expect. It's usually sufficient to wrap your work in an autorelease pool.

If you need further assistance in tracking down what's causing this, please feel free to reach out to us at help@realm.io.

**Comment by luongvanphuc** [comment]
Thanks for your explanation. Sorry that I didn't notice this section on Realm documentation. After reading carefully and doing some test, I finally found the cause and want to let you know.
My simple app doesn't use multi-threading, but because while the app is running (also writing to default.realm), I also use Realm browser to monitor the database, which causes "blocking the thread for a long time" as you said. Please correct me if I'm wrong.

**Comment by segiddins** [comment]
@luongvanphuc that's a possible explanation, and I believe @TimOliver was investigate whether the Browser could actually be causing the version pinning problem.

**Comment by jpsim** [comment]
Yes, the Realm Browser is holding an open read transaction, but only on the main thread, and thus on a runloop, so I don't think (in theory) that it's the browser causing this issue. Although it's certainly worth investigating! @TimOliver can you look into this, please?

**Comment by luongvanphuc** [comment]
I've just done a simple test on your Example source codes (target Simple). First I used Realm Browser (Version 0.91.1) to open the default.realm. Then I created a for loop that inserts 1000 dogs into person (just wrap your code with for loop). As the result, I noticed that the default.realm reached ~60MB.

**Comment by TimOliver** [comment]
Hi everyone! Sorry for the delay.

I've been testing this out myself in order to try and track it down. I ran a couple of sample apps (One that updated existing data, and one that inserted new data) and I got the same result with my insertions app (The .realm file went from 6KB to 26MB and hasn't decreased since). I've been going over the Browser's behavior when it receives an update notification to see if it's doing anything that might be causing a lock-up, but so far, I haven't managed to find anything obvious in there that could be causing this.

I was hoping to have a better response for this before responding. I'll keep on looking!
